go/vt/tabletserver/query_executor.go

```diff
@@ -466,12 +472,40 @@ func (qre *QueryExecutor) execInsertSubquery(conn poolConn) (*mproto.QueryResult
 }
 
 func (qre *QueryExecutor) execInsertPKRows(conn poolConn, pkRows [][]sqltypes.Value) (*mproto.QueryResult, error) {
-	secondaryList, err := buildSecondaryList(qre.plan.TableInfo, pkRows, qre.plan.SecondaryPKValues, qre.bindVars)
+	bsc := buildStreamComment(qre.plan.TableInfo, pkRows, nil)
+	return qre.directFetch(conn, qre.plan.OuterQuery, qre.bindVars, bsc)
+}
+
+func (qre *QueryExecutor) execUpsertPK(conn poolConn, invalidator CacheInvalidator) (*mproto.QueryResult, error) {
+	pkRows, err := buildValueList(qre.plan.TableInfo, qre.plan.PKValues, qre.bindVars)
 	if err != nil {
 		return nil, err
 	}
-	bsc := buildStreamComment(qre.plan.TableInfo, pkRows, secondaryList)
-	return qre.directFetch(conn, qre.plan.OuterQuery, qre.bindVars, bsc)
+	bsc := buildStreamComment(qre.plan.TableInfo, pkRows, nil)
+	result, err := qre.directFetch(conn, qre.plan.OuterQuery, qre.bindVars, bsc)
+	if err == nil {
+		return result, nil
+	}
+	terr, ok := err.(*TabletError)
+	if !ok {
+		return result, err
+	}
+	if terr.SqlError != mysql.ErrDupEntry {
+		return nil, err
+	}
+	// If the error didn't match pk, just return the error without updating.
+	if !strings.Contains(terr.Message, "'PRIMARY'") {
+		return nil, err
+	}
+	result, err = qre.execDMLPKRows(conn, qre.plan.UpsertQuery, pkRows, invalidator)
```

Comment:
So here we get a duplicate key error and decide to ignore the insert error and try an update instead. I think it may worth leaving a comment.